Userland/Utilities/drain.cpp

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2023, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibCore/ArgsParser.h>
+#include <LibCore/File.h>
+#include <LibCore/System.h>
+#include <LibFileSystem/FileSystem.h>
+#include <LibMain/Main.h>
+
+constexpr size_t BLOCK_SIZE = 16 * 16384;
+
+static ErrorOr<void> seek_and_read(size_t offset, ByteBuffer& buffer, Core::File& file)
+{
+    TRY(file.seek(offset, SeekMode::SetPosition));
+    TRY(file.read_some(buffer));
+    return {};
+}
+
+static ErrorOr<void> seek_and_write(size_t offset, ByteBuffer& buffer, Core::File& file)
+{
+    TRY(file.seek(offset, SeekMode::SetPosition));
+    TRY(file.write_some(buffer));
+    return {};
+}
+
+static ErrorOr<void> process_file(Core::File& file, size_t file_size, size_t file_size_rounded)
+{
+    size_t head = 0;
+    size_t tail = file_size_rounded;
+
+    auto head_buffer = TRY(ByteBuffer::create_uninitialized(BLOCK_SIZE));
+    auto tail_buffer = TRY(ByteBuffer::create_uninitialized(BLOCK_SIZE));
+
+    auto stdout = TRY(Core::File::standard_output());
+
+    // Overwrite the current block (after saving its contents to a temporary buffer) with the last block until we've processed half of the blocks in the file.
+    while (head <= tail) {
+        TRY(seek_and_read(head, head_buffer, file));
+        TRY(seek_and_read(tail, tail_buffer, file));
+
+        TRY(seek_and_write(head, tail_buffer, file));
+        TRY(file.truncate(tail));
+
+        TRY(stdout->write_some(head_buffer));
+
+        head += BLOCK_SIZE;
+        tail -= BLOCK_SIZE;
+    };
+
+    size_t remaining_size = file_size - head;
+
+    // Note that we iterate downwards from the end of the file, as the above algorithm left all of the remaining blocks in reverse order.
+    while (remaining_size) {
+        // Account for the fact that the last block may be smaller than BLOCK_SIZE.
+        size_t to_write = remaining_size >= BLOCK_SIZE ? BLOCK_SIZE : BLOCK_SIZE - (file_size_rounded - file_size);
+
+        TRY(seek_and_read(tail, tail_buffer, file));
+        TRY(tail_buffer.try_resize(to_write));
```

Comment:
This will never increase the size, so
```suggestion
        tail_buffer.trim(to_write, true);
```

Reply:
Huh.
Feel free to call this resolved as-is or make that public, your choice.